### PR TITLE
issue/1408 Refactor and Add support for deleting failed datasets

### PR DESCRIPTION
**Pull request by tangrammer**
- [ ] **Update release notes if necessary**

### CODE REVIEW

**Comment by tangrammer** [comment]
@kardan, in backend part a bit of refactoring. You can follow the commits titles to understand better the changes
@finnfiddle, in client-side just update code to support the deleting of failing datasets

File: client/src/actions/dataset.js [hunk 1]
```diff
@@ -302,18 +302,43 @@ function deletePendingDatasetFailure(id, error) {
...
   };
 }
 
+export function deleteFailedDataset(id) {
+  return (dispatch) => {
+    dispatch(deleteDatasetRequest(id));
+    api
+      .del(`/api/data-source/job-execution/${id}/status/failed`)
+      .then(response => response.json())
+      .then(() => dispatch(deleteFailedDatasetSuccess(id)))
```
Comment (finnfiddle) [inline client/src/actions/dataset.js]: you can just use `dispatch(removeDataset(id))` here

File: client/src/actions/dataset.js [hunk 2]
```diff
@@ -302,18 +302,43 @@ function deletePendingDatasetFailure(id, error) {
     error,
   };
 }
+function deleteFailedDatasetSuccess(id) {
```
Comment (finnfiddle) [inline client/src/actions/dataset.js]: this is not needed

Comment (tangrammer) [reply client/src/actions/dataset.js]: Fixed!

File: client/src/actions/dataset.js (hunk 1 above)
Comment (tangrammer) [reply client/src/actions/dataset.js]: Fixed!

**Pull request merged**